crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -783,64 +790,124 @@ where
-                block_import_notification.block_number,
-            )
-            .await?;
-        }
+                mut archiver,
+                older_archived_segments,
+                best_archived_block: (mut best_archived_block_hash, mut best_archived_block_number),
+            } = archiver;
+
+            debug!(%best_archived_block_number, ?best_archived_block_hash,  "Archiver initialized.");
+
+            if let Some(block_import_notification) = saved_block_import_notification.take() {
+                let success = archive_block(
+                    &mut archiver,
+                    segment_headers_store.clone(),
+                    client.clone(),
+                    &sync_oracle,
+                    telemetry.clone(),
+                    archived_segment_notification_sender.clone(),
+                    confirmation_depth_k,
+                    &mut best_archived_block_number,
+                    &mut best_archived_block_hash,
+                    block_import_notification.block_number,
+                )
+                .await?;
 
-        Ok(())
+                if !success {
+                    let error = format!(
+                        "Failed to recover the archiver for block number: {}",
+                        block_import_notification.block_number
+                    );
+                    error!(error);
+
+                    return Err(sp_blockchain::Error::Consensus(sp_consensus::Error::Other(
+                        error.into(),
+                    )));
+                } else {
+                    debug!(
+                        "Saved block import notification succeeded: #{}",
+                        block_import_notification.block_number
+                    );
+                }
+            }
+
+            // Farmers may have not received all previous segments, send them now.
+            for archived_segment in older_archived_segments {
+                send_archived_segment_notification(
+                    &archived_segment_notification_sender,
+                    archived_segment,
+                )
+                .await;
+            }
+
+            while let Some(block_import_notification) =
+                block_importing_notification_stream.next().await
+            {
+                let success = archive_block(
+                    &mut archiver,
+                    segment_headers_store.clone(),
+                    client.clone(),
+                    &sync_oracle,
+                    telemetry.clone(),
+                    archived_segment_notification_sender.clone(),
+                    confirmation_depth_k,
+                    &mut best_archived_block_number,
+                    &mut best_archived_block_hash,
+                    block_import_notification.block_number,
+                )
+                .await?;
+
+                if !success {
+                    warn!(
+                        "Archiver detected an error. \
```

Comment:
It is less of a straight error and more of a controller situation where there is a jump in block number increase. I'd probably make it an `info` level with corresponding messaging rather than a warning. Warning typically means something isn't quite right and needs fixing, but this is an expected situation for us with fast sync.

Reply:
I would make it `debug` then. I initially set it to `warn` because, in theory, the archiver can detect an error outside of the fast sync.

Reply:
With the other suggestions instead of detecting it would not be an error, but expected behavior with recovery expected after some number of attempts. We will still get an error when trying to import a block that contained unexpected segment headers later (in case archiver wasn't initialized and didn't produce segment header or produced incorrect one).

Reply:
The level is debug now, which is good, but the message should change as well because this is not an error as such, this is fully expected and gracefully handled behavior now.